### ISSUE TITLE
Fix all warnings in all repositories

### ISSUE THREAD

**Issue by AntonioND** [blocksds/sdk#4]
To improve code quality, all libraries and tools should be built with ``-Wall -Wextra`` and no warning should be disabled.

This doesn't need to be done in one go, it can be done in little steps if needed.

**Comment by AntonioND** [comment]
All warnings in ``libnds`` have been fixed now.

**Comment by Lorenzooone** [comment]
When compiling, I also use -Wpedantic and -Wstrict-prototypes, and I have noticed warnings because of those.

**Comment by AntonioND** [comment]
https://github.com/blocksds/libnds/commit/b2c8c084a5dd1a40a5a0ea0299028bd52bd5155e

Other libraries, and specially the tools, are in a worse shape.

**Comment by Lorenzooone** [comment]
It did improve significantly. Something I noticed is that  some copy/decompress routines don't have the const qualifier. While I know this is not a real issue, having the qualifiers can be useful...

**Comment by asiekierka** [comment]
Not strictly fixed. See:

```
  CC       src/counter.c
In file included from /home/asie/toolchains/blocksds/sdk/libs/libnds/include/nds/arm9/boxtest.h:19,
                 from /home/asie/toolchains/blocksds/sdk/libs/libnds/include/nds.h:134,
                 from src/compat.h:91,
                 from src/board.h:23,
                 from src/counter.c:32:
/home/asie/toolchains/blocksds/sdk/libs/libnds/include/nds/arm9/videoGL.h: In function 'glCallList':
/home/asie/toolchains/blocksds/sdk/libs/libnds/include/nds/arm9/videoGL.h:856:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  856 |     u32 count = *list++;
      |     ^~~
```

We need to test with the most pedantic warnings for the most restrictive C standard, probably.

**Comment by profi200** [comment]
Just my 2 cents here but i personally find it's sufficient to fix all warnings for -Wall -Wextra. -Wpedantic can be a bit too strict sometimes or give nonsense warnings. That said it doesn't hurt to test -Wpedantic and see what it comes up with but i would not enable it permanently.

**Comment by AntonioND** [comment]
C90 is over 30 years old, I don't think we should care about it, specially when we use our own toolchain. And it's very easy to switch to a more modern standard by using the right compiler flag. 😓 What will happen if we decide to use other things like ``#embed`` when it's finally available? I've never liked that idea of sticking to ancient versions of something for the sake of it. libnds isn't supposed to be built with 20 year old compilers.

This specific GitHub issue is mostly about removing all the flags I've added to tools and libraries that disable specific warnings that are enabled by default with ``-Wall -Wextra``.

**Comment by profi200** [comment]
Just wanted to mention that i think -Wpedantic is more of a guidance. It can't detect all non-standard code anyway. And yeah, i have just switched to C23 for my lib. There are a number of handy improvements. Unfortunately neither gcc nor clang support all of it yet. The long awaited "#embed" isn't supported by any compiler yet.

**Comment by AntonioND** [comment]
Yeah, C23 is promising but it really isn't supported anywhere. Let's hope it happens soon.

And regarding ``-Wpedantic``, I think it's still worth using it as long as you don't butcher your codebase to make the warnings go away. Warnings are there to warn you about potential problems, after all, so any help is good. Something like "you're assigning a const pointer to a non-const pointer" is certainly useful because it may uncover some problem in your logic. However, something I don't consider useful is "a 30 year old version of this language wants you to declare the variable in a different line than the one where you use it". That doesn't actually solve any issue, it's just noise.

**Comment by AntonioND** [comment]
All the warnings of DSWiFi have been fixed now: https://github.com/blocksds/dswifi/commit/dc13c5b5ce89cb7cf3b43c186d1f8246aa0597d2 https://github.com/blocksds/dswifi/commit/a785d9f4204fc34226fb0d71f601f0a37977d232 https://github.com/blocksds/dswifi/commit/f95e4f2124fd6f6508571fc340982a8614af0836

**Comment by AntonioND** [comment]
After version 1.14.1 https://github.com/blocksds/sdk/releases/tag/v1.14.1 the only two repositories that still have a significant number of warnings silenced are ndstool and grit.

**Comment by AntonioND** [comment]
https://codeberg.org/blocksds/sdk/issues/4
